Malformed DNS message on MacOS

### Describe the bug

When starting on macOS this message is printed:
> Can not find {} in the classpath, fallback to system defaults. This may result in incorrect DNS resolutions on MacOS.

Notice the `{}` in the final message.

### Expected behavior

Message tells me what can't be found.

### Actual behavior

☝️

### How to Reproduce?

Start an application in dev mode on macOS.

### Output of `uname -a` or `ver`

macOS 12

### Output of `java -version`

OpenJDK 17.0.2

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.12.0.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 7.4.2

### Additional information

_No response_

I am facing what I think a related issue here. Apart from the WARN message, mongoDB connection fails due to DNS issue... 

This happens right after upgrading from 2.11.2.Final to 2.12.0.Final.


```java
2022-08-31 15:12:46,849 WARN  [io.net.res.dns.DnsServerAddressStreamProviders] (build-52) Can not find {} in the classpath, fallback to system defaults. This may result in incorrect DNS resolutions on MacOS.
[INFO] Checking for existing resources in: /Users/mazen/Desktop/elastico/elastico-platform/src/main/kubernetes.
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2022-08-31 15:12:47,940 INFO  [io.quarkus] (Quarkus Main Thread) platform 1.0.0-SNAPSHOT on JVM (powered by Quarkus 2.12.0.Final) started in 1.642s. Listening on: http://localhost:8080
2022-08-31 15:12:47,940 INFO  [io.quarkus] (Quarkus Main Thread) Profile dev activated. Live Coding activated.
2022-08-31 15:12:47,941 INFO  [io.quarkus] (Quarkus Main Thread) Installed features: [cache, cdi, hibernate-validator, kubernetes, micrometer, mongodb-client, oidc, resteasy-reactive, resteasy-reactive-jackson, security, smallrye-context-propagation, smallrye-health, smallrye-openapi, swagger-ui, vertx]
2022-08-31 15:13:06,273 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (vert.x-eventloop-thread-1) HTTP Request to / failed, error id: 23287a91-dbad-4d42-85db-5f074c41ec6e-1: io.quarkus.cache.CacheException: com.mongodb.MongoConfigurationException: Failed looking up TXT record for host cluster0.3olzb.mongodb.net
	at io.quarkus.cache.runtime.CacheResultInterceptor$1.apply(CacheResultInterceptor.java:78)
	at io.quarkus.cache.runtime.CacheResultInterceptor$1.apply(CacheResultInterceptor.java:63)
	at io.smallrye.context.impl.wrappers.SlowContextualFunction.apply(SlowContextualFunction.java:21)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.performInnerSubscription(UniOnItemTransformToUni.java:68)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.onItem(UniOnItemTransformToUni.java:57)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromCompletionStage$CompletionStageUniSubscription.forwardResult(UniCreateFromCompletionStage.java:63)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:887)
	at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2325)
	at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:144)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromCompletionStage$CompletionStageUniSubscription.forward(UniCreateFromCompletionStage.java:51)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromCompletionStage.subscribe(UniCreateFromCompletionStage.java:35)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni.subscribe(UniOnItemTransformToUni.java:25)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni.subscribe(UniOnItemTransformToUni.java:25)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.performInnerSubscription(UniOnItemTransformToUni.java:81)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.onItem(UniOnItemTransformToUni.java:57)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromKnownItem$KnownItemSubscription.forward(UniCreateFromKnownItem.java:38)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromKnownItem$KnownItemSubscription.access$100(UniCreateFromKnownItem.java:26)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromKnownItem.subscribe(UniCreateFromKnownItem.java:23)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni.subscribe(UniOnItemTransformToUni.java:25)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransform.subscribe(UniOnItemTransform.java:22)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni.subscribe(UniOnItemTransformToUni.java:25)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.performInnerSubscription(UniOnItemTransformToUni.java:81)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.onItem(UniOnItemTransformToUni.java:57)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromCompletionStage$CompletionStageUniSubscription.forwardResult(UniCreateFromCompletionStage.java:63)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:887)
	at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2325)
	at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:144)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromCompletionStage$CompletionStageUniSubscription.forward(UniCreateFromCompletionStage.java:51)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromCompletionStage.subscribe(UniCreateFromCompletionStage.java:35)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni.subscribe(UniOnItemTransformToUni.java:25)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransform.subscribe(UniOnItemTransform.java:22)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniMemoizeOp.subscribe(UniMemoizeOp.java:84)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransform.subscribe(UniOnItemTransform.java:22)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransform.subscribe(UniOnItemTransform.java:22)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransform.subscribe(UniOnItemTransform.java:22)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni.subscribe(UniOnItemTransformToUni.java:25)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniMemoizeOp.subscribe(UniMemoizeOp.java:84)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.groups.UniSubscribe.withSubscriber(UniSubscribe.java:52)
	at io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$2.handle(HttpSecurityRecorder.java:113)
	at io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$2.handle(HttpSecurityRecorder.java:60)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:140)
	at io.quarkus.vertx.http.runtime.filters.accesslog.AccessLogHandler.handle(AccessLogHandler.java:151)
	at io.quarkus.vertx.http.runtime.filters.accesslog.AccessLogHandler.handle(AccessLogHandler.java:93)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:140)
	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$5.handle(VertxHttpHotReplacementSetup.java:196)
	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$5.handle(VertxHttpHotReplacementSetup.java:185)
	at io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:141)
	at io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:54)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:503)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: com.mongodb.MongoConfigurationException: Failed looking up TXT record for host cluster0.3olzb.mongodb.net
	at com.mongodb.internal.dns.DefaultDnsResolver.resolveAdditionalQueryParametersFromTxtRecords(DefaultDnsResolver.java:133)
	at com.mongodb.ConnectionString.<init>(ConnectionString.java:397)
	at io.quarkus.mongodb.runtime.MongoClients.createMongoConfiguration(MongoClients.java:262)
	at io.quarkus.mongodb.runtime.MongoClients.createReactiveMongoClient(MongoClients.java:126)
	at io.quarkus.mongodb.runtime.MongoClients_Subclass.createReactiveMongoClient$$superforward1(Unknown Source)
	at io.quarkus.mongodb.runtime.MongoClients_Subclass$$function$$5.apply(Unknown Source)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:53)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.proceed(InvocationInterceptor.java:62)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.monitor(InvocationInterceptor.java:51)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor_Bean.intercept(Unknown Source)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:40)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
	at io.quarkus.mongodb.runtime.MongoClients_Subclass.createReactiveMongoClient(Unknown Source)
	at io.quarkus.mongodb.runtime.MongoClientRecorder.lambda$reactiveMongoClientSupplier$1(MongoClientRecorder.java:68)
	at io.quarkus.mongodb.runtime.MongoClientRecorder$MongoClientSupplier.get(MongoClientRecorder.java:57)
	at io.quarkus.mongodb.reactive.ReactiveMongoClient_afcbcf6dad636d255df730977b21f823ec3c9f2e_Synthetic_Bean.create(Unknown Source)
	at io.quarkus.mongodb.reactive.ReactiveMongoClient_afcbcf6dad636d255df730977b21f823ec3c9f2e_Synthetic_Bean.create(Unknown Source)
	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:111)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:35)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:32)
	at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:26)
	at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:32)
	at io.quarkus.arc.impl.ClientProxies.getApplicationScopedDelegate(ClientProxies.java:19)
	at io.quarkus.mongodb.reactive.ReactiveMongoClient_afcbcf6dad636d255df730977b21f823ec3c9f2e_Synthetic_ClientProxy.arc$delegate(Unknown Source)
	at io.quarkus.mongodb.reactive.ReactiveMongoClient_afcbcf6dad636d255df730977b21f823ec3c9f2e_Synthetic_ClientProxy.getDatabase(Unknown Source)
	at com.elastico.platform.dao.nosql.mongodb.MongoDao.getDatabase(MongoDao.java:229)
	at com.elastico.platform.dao.nosql.mongodb.MongoDao.getCollection(MongoDao.java:225)
	at com.elastico.platform.dao.nosql.mongodb.MongoDao.selectWhere(MongoDao.java:66)
	at com.elastico.platform.metadata.definition.MetadataDefinitionManager.createMetadataDefinition(MetadataDefinitionManager.java:44)
	at com.elastico.platform.metadata.definition.MetadataDefinitionManager.getMetadataDefinitionByDesignName(MetadataDefinitionManager.java:32)
	at com.elastico.platform.metadata.definition.MetadataDefinitionManager_Subclass.getMetadataDefinitionByDesignName$$superforward1(Unknown Source)
	at com.elastico.platform.metadata.definition.MetadataDefinitionManager_Subclass$$function$$4.apply(Unknown Source)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:53)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.proceed(InvocationInterceptor.java:62)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.monitor(InvocationInterceptor.java:51)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor_Bean.intercept(Unknown Source)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:49)
	at io.quarkus.cache.runtime.CacheResultInterceptor$1.apply(CacheResultInterceptor.java:68)
	... 77 more
Caused by: com.mongodb.MongoConfigurationException: Unable to look up TXT record for host cluster0.3olzb.mongodb.net
	at io.quarkus.mongodb.runtime.dns.MongoDnsClient.resolveTxtRequest(MongoDnsClient.java:152)
	at io.quarkus.mongodb.runtime.dns.MongoDnsClient.getResourceRecordData(MongoDnsClient.java:98)
	at com.mongodb.internal.dns.DefaultDnsResolver.resolveAdditionalQueryParametersFromTxtRecords(DefaultDnsResolver.java:114)
	... 117 more
Caused by: java.lang.IllegalStateException: The current thread cannot be blocked: vert.x-eventloop-thread-1
	at io.smallrye.mutiny.operators.uni.UniBlockingAwait.await(UniBlockingAwait.java:30)
	at io.smallrye.mutiny.groups.UniAwait.atMost(UniAwait.java:65)
	at io.quarkus.mongodb.runtime.dns.MongoDnsClient.resolveTxtRequest(MongoDnsClient.java:150)
	... 119 more
```

--- a/extensions/netty/deployment/src/main/java/io/quarkus/netty/deployment/JBossNettyLoggerFactory.java
+++ b/extensions/netty/deployment/src/main/java/io/quarkus/netty/deployment/JBossNettyLoggerFactory.java
@@ -1,5 +1,9 @@
 package io.quarkus.netty.deployment;
 
+import java.text.MessageFormat;
+import java.util.HashSet;
+import java.util.Set;
+
 import org.jboss.logging.Logger;
 
 import io.netty.util.internal.logging.AbstractInternalLogger;
@@ -29,26 +33,43 @@ public class JBossNettyLoggerFactory extends InternalLoggerFactory {
 
         @Override
         public void trace(String msg) {
-            log.trace(msg);
+            if (isTraceEnabled()) {
+                log.trace(msg);
+            }
         }
 
         @Override
         public void trace(String format, Object arg) {
-            log.tracef(format, arg);
+            if (isTraceEnabled()) {
+                FormattingTuple ft = MessageFormatter.format(format, arg);
+                trace0(ft.getMessage(), ft.getThrowable());
+            }
         }
 
         @Override
         public void trace(String format, Object argA, Object argB) {
-            log.tracef(format, argA, argB);
+            if (isTraceEnabled()) {
+                FormattingTuple ft = MessageFormatter.format(format, argA, argB);
+                trace0(ft.getMessage(), ft.getThrowable());
+            }
         }
 
         @Override
         public void trace(String format, Object... arguments) {
-            log.tracef(format, arguments);
+            if (isTraceEnabled()) {
+                FormattingTuple ft = MessageFormatter.format(format, arguments);
+                trace0(ft.getMessage(), ft.getThrowable());
+            }
         }
 
         @Override
         public void trace(String msg, Throwable t) {
+            if (isTraceEnabled()) {
+                trace0(msg, t);
+            }
+        }
+
+        private void trace0(String msg, Throwable t) {
             log.trace(msg, t);
         }
 
@@ -59,26 +80,43 @@ public class JBossNettyLoggerFactory extends InternalLoggerFactory {
 
         @Override
         public void debug(String msg) {
-            log.debug(msg);
+            if (isDebugEnabled()) {
+                log.debug(msg);
+            }
         }
 
         @Override
         public void debug(String format, Object arg) {
-            log.debugf(format, arg);
+            if (isDebugEnabled()) {
+                FormattingTuple ft = MessageFormatter.format(format, arg);
+                debug0(ft.getMessage(), ft.getThrowable());
+            }
         }
 
         @Override
         public void debug(String format, Object argA, Object argB) {
-            log.debugf(format, argA, argB);
+            if (isDebugEnabled()) {
+                FormattingTuple ft = MessageFormatter.format(format, argA, argB);
+                debug0(ft.getMessage(), ft.getThrowable());
+            }
         }
 
         @Override
         public void debug(String format, Object... arguments) {
-            log.debugf(format, arguments);
+            if (isDebugEnabled()) {
+                FormattingTuple ft = MessageFormatter.format(format, arguments);
+                debug0(ft.getMessage(), ft.getThrowable());
+            }
         }
 
         @Override
         public void debug(String msg, Throwable t) {
+            if (isDebugEnabled()) {
+                debug0(msg, t);
+            }
+        }
+
+        private void debug0(String msg, Throwable t) {
             log.debug(msg, t);
         }
 
@@ -89,26 +127,43 @@ public class JBossNettyLoggerFactory extends InternalLoggerFactory {
 
         @Override
         public void info(String msg) {
-            log.info(msg);
+            if (isInfoEnabled()) {
+                log.info(msg);
+            }
         }
 
         @Override
         public void info(String format, Object arg) {
-            log.infof(format, arg);
+            if (isInfoEnabled()) {
+                FormattingTuple ft = MessageFormatter.format(format, arg);
+                info0(ft.getMessage(), ft.getThrowable());
+            }
         }
 
         @Override
         public void info(String format, Object argA, Object argB) {
-            log.infof(format, argA, argB);
+            if (isInfoEnabled()) {
+                FormattingTuple ft = MessageFormatter.format(format, argA, argB);
+                info0(ft.getMessage(), ft.getThrowable());
+            }
         }
 
         @Override
         public void info(String format, Object... arguments) {
-            log.infof(format, arguments);
+            if (isInfoEnabled()) {
+                FormattingTuple ft = MessageFormatter.format(format, arguments);
+                info0(ft.getMessage(), ft.getThrowable());
+            }
         }
 
         @Override
         public void info(String msg, Throwable t) {
+            if (isInfoEnabled()) {
+                info0(msg, t);
+            }
+        }
+
+        private void info0(String msg, Throwable t) {
             log.info(msg, t);
         }
 
@@ -119,26 +174,43 @@ public class JBossNettyLoggerFactory extends InternalLoggerFactory {
 
         @Override
         public void warn(String msg) {
-            log.warn(msg);
+            if (isWarnEnabled()) {
+                log.warn(msg);
+            }
         }
 
         @Override
         public void warn(String format, Object arg) {
-            log.warnf(format, arg);
+            if (isWarnEnabled()) {
+                FormattingTuple ft = MessageFormatter.format(format, arg);
+                warn0(ft.getMessage(), ft.getThrowable());
+            }
         }
 
         @Override
         public void warn(String format, Object... arguments) {
-            log.warnf(format, arguments);
+            if (isWarnEnabled()) {
+                FormattingTuple ft = MessageFormatter.format(format, arguments);
+                warn0(ft.getMessage(), ft.getThrowable());
+            }
         }
 
         @Override
         public void warn(String format, Object argA, Object argB) {
-            log.warnf(format, argA, argB);
+            if (isWarnEnabled()) {
+                FormattingTuple ft = MessageFormatter.format(format, argA, argB);
+                warn0(ft.getMessage(), ft.getThrowable());
+            }
         }
 
         @Override
         public void warn(String msg, Throwable t) {
+            if (isWarnEnabled()) {
+                warn0(msg, t);
+            }
+        }
+
+        private void warn0(String msg, Throwable t) {
             log.warn(msg, t);
         }
 
@@ -149,29 +221,466 @@ public class JBossNettyLoggerFactory extends InternalLoggerFactory {
 
         @Override
         public void error(String msg) {
-            log.error(msg);
+            if (isErrorEnabled()) {
+                log.error(msg);
+            }
         }
 
         @Override
         public void error(String format, Object arg) {
-            log.errorf(format, arg);
+            if (isErrorEnabled()) {
+                FormattingTuple ft = MessageFormatter.format(format, arg);
+                error0(ft.getMessage(), ft.getThrowable());
+            }
         }
 
         @Override
         public void error(String format, Object argA, Object argB) {
-            log.errorf(format, argA, argB);
+            if (isErrorEnabled()) {
+                FormattingTuple ft = MessageFormatter.format(format, argA, argB);
+                error0(ft.getMessage(), ft.getThrowable());
+            }
         }
 
         @Override
         public void error(String format, Object... arguments) {
-            log.errorf(format, arguments);
+            if (isErrorEnabled()) {
+                FormattingTuple ft = MessageFormatter.format(format, arguments);
+                error0(ft.getMessage(), ft.getThrowable());
+            }
         }
 
         @Override
         public void error(String msg, Throwable t) {
+            if (isErrorEnabled()) {
+                error0(msg, t);
+            }
+        }
+
+        private void error0(String msg, Throwable t) {
             log.error(msg, t);
         }
 
+    }
+
+    // MessageFormatter class copied from Netty because its methods are package private
+
+    /*
+     * Copyright 2013 The Netty Project
+     *
+     * The Netty Project licenses this file to you under the Apache License,
+     * version 2.0 (the "License"); you may not use this file except in compliance
+     * with the License. You may obtain a copy of the License at:
+     *
+     * https://www.apache.org/licenses/LICENSE-2.0
+     *
+     * Unless required by applicable law or agreed to in writing, software
+     * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+     * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+     * License for the specific language governing permissions and limitations
+     * under the License.
+     */
+
+    /**
+     * Formats messages according to very simple substitution rules. Substitutions
+     * can be made 1, 2 or more arguments.
+     * <p/>
+     * <p/>
+     * For example,
+     * <p/>
+     *
+     * <pre>
+     * MessageFormatter.format(&quot;Hi {}.&quot;, &quot;there&quot;)
+     * </pre>
+     * <p/>
+     * will return the string "Hi there.".
+     * <p/>
+     * The {} pair is called the <em>formatting anchor</em>. It serves to designate
+     * the location where arguments need to be substituted within the message
+     * pattern.
+     * <p/>
+     * In case your message contains the '{' or the '}' character, you do not have
+     * to do anything special unless the '}' character immediately follows '{'. For
+     * example,
+     * <p/>
+     *
+     * <pre>
+     * MessageFormatter.format(&quot;Set {1,2,3} is not equal to {}.&quot;, &quot;1,2&quot;);
+     * </pre>
+     * <p/>
+     * will return the string "Set {1,2,3} is not equal to 1,2.".
+     * <p/>
+     * <p/>
+     * If for whatever reason you need to place the string "{}" in the message
+     * without its <em>formatting anchor</em> meaning, then you need to escape the
+     * '{' character with '\', that is the backslash character. Only the '{'
+     * character should be escaped. There is no need to escape the '}' character.
+     * For example,
+     * <p/>
+     *
+     * <pre>
+     * MessageFormatter.format(&quot;Set \\{} is not equal to {}.&quot;, &quot;1,2&quot;);
+     * </pre>
+     * <p/>
+     * will return the string "Set {} is not equal to 1,2.".
+     * <p/>
+     * <p/>
+     * The escaping behavior just described can be overridden by escaping the escape
+     * character '\'. Calling
+     * <p/>
+     *
+     * <pre>
+     * MessageFormatter.format(&quot;File name is C:\\\\{}.&quot;, &quot;file.zip&quot;);
+     * </pre>
+     * <p/>
+     * will return the string "File name is C:\file.zip".
+     * <p/>
+     * <p/>
+     * The formatting conventions are different than those of {@link MessageFormat}
+     * which ships with the Java platform. This is justified by the fact that
+     * SLF4J's implementation is 10 times faster than that of {@link MessageFormat}.
+     * This local performance difference is both measurable and significant in the
+     * larger context of the complete logging processing chain.
+     * <p/>
+     * <p/>
+     * See also {@link #format(String, Object)},
+     * {@link #format(String, Object, Object)} and
+     * {@link #arrayFormat(String, Object[])} methods for more details.
+     */
+    public static final class MessageFormatter {
+        private static final String DELIM_STR = "{}";
+        private static final char ESCAPE_CHAR = '\\';
+
+        /**
+         * Performs single argument substitution for the 'messagePattern' passed as
+         * parameter.
+         * <p/>
+         * For example,
+         * <p/>
+         *
+         * <pre>
+         * MessageFormatter.format(&quot;Hi {}.&quot;, &quot;there&quot;);
+         * </pre>
+         * <p/>
+         * will return the string "Hi there.".
+         * <p/>
+         *
+         * @param messagePattern The message pattern which will be parsed and formatted
+         * @param arg The argument to be substituted in place of the formatting anchor
+         * @return The formatted message
+         */
+        static FormattingTuple format(String messagePattern, Object arg) {
+            return arrayFormat(messagePattern, new Object[] { arg });
+        }
+
+        /**
+         * Performs a two argument substitution for the 'messagePattern' passed as
+         * parameter.
+         * <p/>
+         * For example,
+         * <p/>
+         *
+         * <pre>
+         * MessageFormatter.format(&quot;Hi {}. My name is {}.&quot;, &quot;Alice&quot;, &quot;Bob&quot;);
+         * </pre>
+         * <p/>
+         * will return the string "Hi Alice. My name is Bob.".
+         *
+         * @param messagePattern The message pattern which will be parsed and formatted
+         * @param argA The argument to be substituted in place of the first formatting
+         *        anchor
+         * @param argB The argument to be substituted in place of the second formatting
+         *        anchor
+         * @return The formatted message
+         */
+        static FormattingTuple format(final String messagePattern,
+                Object argA, Object argB) {
+            return arrayFormat(messagePattern, new Object[] { argA, argB });
+        }
+
+        /**
+         * Same principle as the {@link #format(String, Object)} and
+         * {@link #format(String, Object, Object)} methods except that any number of
+         * arguments can be passed in an array.
+         *
+         * @param messagePattern The message pattern which will be parsed and formatted
+         * @param argArray An array of arguments to be substituted in place of formatting
+         *        anchors
+         * @return The formatted message
+         */
+        static FormattingTuple arrayFormat(final String messagePattern,
+                final Object[] argArray) {
+            if (argArray == null || argArray.length == 0) {
+                return new FormattingTuple(messagePattern, null);
+            }
+
+            int lastArrIdx = argArray.length - 1;
+            Object lastEntry = argArray[lastArrIdx];
+            Throwable throwable = lastEntry instanceof Throwable ? (Throwable) lastEntry : null;
+
+            if (messagePattern == null) {
+                return new FormattingTuple(null, throwable);
+            }
+
+            int j = messagePattern.indexOf(DELIM_STR);
+            if (j == -1) {
+                // this is a simple string
+                return new FormattingTuple(messagePattern, throwable);
+            }
+
+            StringBuilder sbuf = new StringBuilder(messagePattern.length() + 50);
+            int i = 0;
+            int L = 0;
+            do {
+                boolean notEscaped = j == 0 || messagePattern.charAt(j - 1) != ESCAPE_CHAR;
+                if (notEscaped) {
+                    // normal case
+                    sbuf.append(messagePattern, i, j);
+                } else {
+                    sbuf.append(messagePattern, i, j - 1);
+                    // check that escape char is not is escaped: "abc x:\\{}"
+                    notEscaped = j >= 2 && messagePattern.charAt(j - 2) == ESCAPE_CHAR;
+                }
+
+                i = j + 2;
+                if (notEscaped) {
+                    deeplyAppendParameter(sbuf, argArray[L], null);
+                    L++;
+                    if (L > lastArrIdx) {
+                        break;
+                    }
+                } else {
+                    sbuf.append(DELIM_STR);
+                }
+                j = messagePattern.indexOf(DELIM_STR, i);
+            } while (j != -1);
+
+            // append the characters following the last {} pair.
+            sbuf.append(messagePattern, i, messagePattern.length());
+            return new FormattingTuple(sbuf.toString(), L <= lastArrIdx ? throwable : null);
+        }
+
+        // special treatment of array values was suggested by 'lizongbo'
+        private static void deeplyAppendParameter(StringBuilder sbuf, Object o,
+                Set<Object[]> seenSet) {
+            if (o == null) {
+                sbuf.append("null");
+                return;
+            }
+            Class<?> objClass = o.getClass();
+            if (!objClass.isArray()) {
+                if (Number.class.isAssignableFrom(objClass)) {
+                    // Prevent String instantiation for some number types
+                    if (objClass == Long.class) {
+                        sbuf.append(((Long) o).longValue());
+                    } else if (objClass == Integer.class || objClass == Short.class || objClass == Byte.class) {
+                        sbuf.append(((Number) o).intValue());
+                    } else if (objClass == Double.class) {
+                        sbuf.append(((Double) o).doubleValue());
+                    } else if (objClass == Float.class) {
+                        sbuf.append(((Float) o).floatValue());
+                    } else {
+                        safeObjectAppend(sbuf, o);
+                    }
+                } else {
+                    safeObjectAppend(sbuf, o);
+                }
+            } else {
+                // check for primitive array types because they
+                // unfortunately cannot be cast to Object[]
+                sbuf.append('[');
+                if (objClass == boolean[].class) {
+                    booleanArrayAppend(sbuf, (boolean[]) o);
+                } else if (objClass == byte[].class) {
+                    byteArrayAppend(sbuf, (byte[]) o);
+                } else if (objClass == char[].class) {
+                    charArrayAppend(sbuf, (char[]) o);
+                } else if (objClass == short[].class) {
+                    shortArrayAppend(sbuf, (short[]) o);
+                } else if (objClass == int[].class) {
+                    intArrayAppend(sbuf, (int[]) o);
+                } else if (objClass == long[].class) {
+                    longArrayAppend(sbuf, (long[]) o);
+                } else if (objClass == float[].class) {
+                    floatArrayAppend(sbuf, (float[]) o);
+                } else if (objClass == double[].class) {
+                    doubleArrayAppend(sbuf, (double[]) o);
+                } else {
+                    objectArrayAppend(sbuf, (Object[]) o, seenSet);
+                }
+                sbuf.append(']');
+            }
+        }
+
+        private static void safeObjectAppend(StringBuilder sbuf, Object o) {
+            try {
+                String oAsString = o.toString();
+                sbuf.append(oAsString);
+            } catch (Throwable t) {
+                System.err
+                        .println("SLF4J: Failed toString() invocation on an object of type ["
+                                + o.getClass().getName() + ']');
+                t.printStackTrace();
+                sbuf.append("[FAILED toString()]");
+            }
+        }
+
+        private static void objectArrayAppend(StringBuilder sbuf, Object[] a, Set<Object[]> seenSet) {
+            if (a.length == 0) {
+                return;
+            }
+            if (seenSet == null) {
+                seenSet = new HashSet<Object[]>(a.length);
+            }
+            if (seenSet.add(a)) {
+                deeplyAppendParameter(sbuf, a[0], seenSet);
+                for (int i = 1; i < a.length; i++) {
+                    sbuf.append(", ");
+                    deeplyAppendParameter(sbuf, a[i], seenSet);
+                }
+                // allow repeats in siblings
+                seenSet.remove(a);
+            } else {
+                sbuf.append("...");
+            }
+        }
+
+        private static void booleanArrayAppend(StringBuilder sbuf, boolean[] a) {
+            if (a.length == 0) {
+                return;
+            }
+            sbuf.append(a[0]);
+            for (int i = 1; i < a.length; i++) {
+                sbuf.append(", ");
+                sbuf.append(a[i]);
+            }
+        }
+
+        private static void byteArrayAppend(StringBuilder sbuf, byte[] a) {
+            if (a.length == 0) {
+                return;
+            }
+            sbuf.append(a[0]);
+            for (int i = 1; i < a.length; i++) {
+                sbuf.append(", ");
+                sbuf.append(a[i]);
+            }
+        }
+
+        private static void charArrayAppend(StringBuilder sbuf, char[] a) {
+            if (a.length == 0) {
+                return;
+            }
+            sbuf.append(a[0]);
+            for (int i = 1; i < a.length; i++) {
+                sbuf.append(", ");
+                sbuf.append(a[i]);
+            }
+        }
+
+        private static void shortArrayAppend(StringBuilder sbuf, short[] a) {
+            if (a.length == 0) {
+                return;
+            }
+            sbuf.append(a[0]);
+            for (int i = 1; i < a.length; i++) {
+                sbuf.append(", ");
+                sbuf.append(a[i]);
+            }
+        }
+
+        private static void intArrayAppend(StringBuilder sbuf, int[] a) {
+            if (a.length == 0) {
+                return;
+            }
+            sbuf.append(a[0]);
+            for (int i = 1; i < a.length; i++) {
+                sbuf.append(", ");
+                sbuf.append(a[i]);
+            }
+        }
+
+        private static void longArrayAppend(StringBuilder sbuf, long[] a) {
+            if (a.length == 0) {
+                return;
+            }
+            sbuf.append(a[0]);
+            for (int i = 1; i < a.length; i++) {
+                sbuf.append(", ");
+                sbuf.append(a[i]);
+            }
+        }
+
+        private static void floatArrayAppend(StringBuilder sbuf, float[] a) {
+            if (a.length == 0) {
+                return;
+            }
+            sbuf.append(a[0]);
+            for (int i = 1; i < a.length; i++) {
+                sbuf.append(", ");
+                sbuf.append(a[i]);
+            }
+        }
+
+        private static void doubleArrayAppend(StringBuilder sbuf, double[] a) {
+            if (a.length == 0) {
+                return;
+            }
+            sbuf.append(a[0]);
+            for (int i = 1; i < a.length; i++) {
+                sbuf.append(", ");
+                sbuf.append(a[i]);
+            }
+        }
+
+        private MessageFormatter() {
+        }
+    }
+
+    // FormattingTuple class copied from Netty because its constructor is package private
+
+    /**
+     * Copyright (c) 2004-2011 QOS.ch
+     * All rights reserved.
+     *
+     * Permission is hereby granted, free of charge, to any person obtaining
+     * a copy of this software and associated documentation files (the
+     * "Software"), to deal in the Software without restriction, including
+     * without limitation the rights to use, copy, modify, merge, publish,
+     * distribute, sublicense, and/or sell copies of the Software, and to
+     * permit persons to whom the Software is furnished to do so, subject to
+     * the following conditions:
+     *
+     * The above copyright notice and this permission notice shall be
+     * included in all copies or substantial portions of the Software.
+     *
+     * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+     * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+     * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+     * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+     * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+     * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+     * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+     *
+     */
+
+    public static final class FormattingTuple {
+
+        private final String message;
+        private final Throwable throwable;
+
+        FormattingTuple(String message, Throwable throwable) {
+            this.message = message;
+            this.throwable = throwable;
+        }
+
+        public String getMessage() {
+            return message;
+        }
+
+        public Throwable getThrowable() {
+            return throwable;
+        }
     }
 
 }

